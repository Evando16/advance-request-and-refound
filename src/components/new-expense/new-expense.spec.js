import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import pretty from 'pretty';

import NewExpense from './new-expense';

jest.mock('./new-expense-service.js');

describe('NewExpenseComponent', () => {
  const fileMock = {
    file: {
      lastModified: 1588893767728,
      lastModifiedDate: new Date('2020-10-10'),
      name: 'image_test.png',
      size: 377755,
      type: 'image/png',
      webkitRelativePath: '',
    },
    error: '',
  };

  const emptyExpenseMock = {
    expenseType: '',
    currency: '',
    description: '',
    receiptDate: '',
    receiptValue: '',
    valueToBePaid: '',
    receiptImage: {},
  };

  const setExpense = jest.fn();
  const onSubmit = jest.fn((event) => event.preventDefault());

  let component = null;

  describe('Component tests', () => {
    beforeEach(() => {
      component = render(
        <NewExpense expense={emptyExpenseMock} setExpense={setExpense} onSubmit={onSubmit} />,
      );
    });
    it('should render receipt value field when chose currency', () => {
      component = render(
        <NewExpense
          expense={{ ...emptyExpenseMock, currency: 'BRL' }}
          setExpense={setExpense}
          onSubmit={onSubmit}
        />,
      );

      const receiptValue = component.getByTestId('new-expense__receipt-value');
      expect(receiptValue).not.toBeNull();
    });

    it('should render to be paid field when chose currency', () => {
      component = render(
        <NewExpense
          expense={{ ...emptyExpenseMock, currency: 'BRL' }}
          setExpense={setExpense}
          onSubmit={onSubmit}
        />,
      );

      const receiptValue = component.getByTestId('new-expense__value-to-be-paid');
      expect(receiptValue).not.toBeNull();
    });

    it('should set new expense on selecting file ', () => {
      const fileInput = component.getByTestId('new-expense__file');
      fireEvent.change(fileInput, { target: { files: [fileMock.file] } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, receiptImage: fileMock });
    });

    it('should file erro when file is not a .png or .jpg ', () => {
      const invalidFile = { ...fileMock.file, name: 'invalid.txt' };
      const fileInput = component.getByTestId('new-expense__file');
      fireEvent.change(fileInput, { target: { files: [invalidFile] } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, receiptImage: { error: 'Arquivo deve ser .jpg ou .png', file: null } });
    });

    it('should file erro when file is not a .png or .jpg ', () => {
      const invalidFile = { ...fileMock.file, size: 2048576 };
      const fileInput = component.getByTestId('new-expense__file');
      fireEvent.change(fileInput, { target: { files: [invalidFile] } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, receiptImage: { error: 'Tamanho da imagem deve ser menor que 1MB', file: null } });
    });

    it('should set new expense on change expense type ', () => {
      const typeInput = component.getByTestId('new-expense__type-select');
      fireEvent.change(typeInput, { target: { value: 'hotel-fee' } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, expenseType: 'hotel-fee' });
    });

    it('should set new expense on change description ', () => {
      const descriptionInput = component.getByTestId('new-expense__description');
      fireEvent.change(descriptionInput, { target: { value: 'description value' } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, description: 'description value' });
    });

    it('should set new expense on change receipt date ', () => {
      const receiptDateInput = component.getByTestId('new-expense__receipt-date');
      fireEvent.change(receiptDateInput, { target: { value: '2020-01-10' } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, receiptDate: '2020-01-10' });
    });

    it('should set new expense on change currency ', () => {
      const currencyInput = component.getByTestId('new-expense__currency-select');
      fireEvent.change(currencyInput, { target: { value: 'BRL' } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, currency: 'BRL' });
    });

    it('should set new expense on change receipt value ', () => {
      component = render(
        <NewExpense
          expense={{ ...emptyExpenseMock, currency: 'BRL' }}
          setExpense={setExpense}
          onSubmit={onSubmit}
        />,
      );

      const receiptValueInput = component.getByTestId('new-expense__receipt-value');
      fireEvent.change(receiptValueInput, { target: { value: '1616.16' } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, currency: 'BRL', receiptValue: '1616.16' });
    });

    it('should set new expense on change value to be paid', () => {
      component = render(
        <NewExpense
          expense={{ ...emptyExpenseMock, currency: 'BRL' }}
          setExpense={setExpense}
          onSubmit={onSubmit}
        />,
      );

      const valueToBePaidInput = component.getByTestId('new-expense__value-to-be-paid');
      fireEvent.change(valueToBePaidInput, { target: { value: '100.16' } });

      expect(setExpense).toHaveBeenCalledWith({ ...emptyExpenseMock, currency: 'BRL', valueToBePaid: '100.16' });
    });

    it('should save new expense', () => {
      const btn = component.getByTestId('new-expense__btn-submit');
      expect(onSubmit).not.toHaveBeenCalled();
      fireEvent.click(btn);

      expect(onSubmit).toHaveBeenCalled();
    });
  });

  describe('Snapshot tests', () => {
    it('should render new expense component without conditional fields', () => {
      const expenseMock = {
        expenseType: 'food',
        currency: '',
        description: 'description',
        receiptDate: '2020-10-10',
        receiptValue: '',
        valueToBePaid: '',
        receiptImage: fileMock,
      };
      component = render(
        <NewExpense expense={expenseMock} setExpense={setExpense} onSubmit={onSubmit} />,
      );

      expect(pretty(component.container.innerHTML)).toMatchSnapshot();
    });

    it('should render new expense component with conditional fields', () => {
      const expenseMock = {
        expenseType: 'food',
        currency: 'BRL',
        description: 'description',
        receiptDate: '2020-10-10',
        receiptValue: '16',
        valueToBePaid: '10',
        receiptImage: fileMock,
      };
      component = render(
        <NewExpense expense={expenseMock} setExpense={setExpense} onSubmit={onSubmit} />,
      );

      const input = component.getByTestId('new-expense__currency-select');
      fireEvent.change(input, { target: { value: 'BRL' } });

      expect(pretty(component.container.innerHTML)).toMatchSnapshot();
    });
  });
});
