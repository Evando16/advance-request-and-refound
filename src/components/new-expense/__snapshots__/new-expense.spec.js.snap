// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewExpenseComponent Snapshot tests should render new expense component with conditional fields 1`] = `
"<div class=\\"NewExpenseContainer-hqtosb juPdFW\\">
  <form class=\\"NewExpenseForm-cwuXvK kEHXvR\\"><span class=\\"NewExpenseTitle-lhcAMw jEIPWO\\">New expense</span>
    <div class=\\"NewExpenseContent-jatTpR kRaxfX\\">
      <div class=\\"NewExpenseFileContainer-bNEBXl cdSBvg\\">
        <div class=\\"NewExpenseFile-kZrDhS ldnNcp\\"><span class=\\"NewExpenseSpanFieldTitle-fryOLC eNcQVo\\">Receipt or invoice*</span><label for=\\"new-expense__file\\" class=\\"NewExpenseFileButton-doqqYw dmgrQy\\"><span>Select file</span><input id=\\"new-expense__file\\" data-testid=\\"new-expense__file\\" style=\\"display: none;\\" accept=\\".jpg,.png\\" type=\\"file\\"></label><span class=\\"NewExpenseHint-gvuMzb hbrupz\\">Just JPG or PNG images are accepted</span><span color=\\"#f44336\\" class=\\"NewExpenseHint-gvuMzb kyTpiy\\">error on file</span></div>
      </div>
      <div class=\\"NewExpenseFieldsContainer-eXgEkW kWXdsa\\"><label for=\\"new-expense__type-selec\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Type*</span><select id=\\"new-expense__type-select\\" data-testid=\\"new-expense__type-select\\" required=\\"\\" class=\\"NewExpenseSelectField-jVTKdO iFJTld\\">
            <option value=\\"\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\"></option>
            <option value=\\"hotel-fee\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">Hotel</option>
            <option value=\\"food\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">Food</option>
            <option value=\\"transport\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">Transport</option>
          </select></label><label for=\\"new-expense__description\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Expense description*</span><input required=\\"\\" id=\\"new-expense__description\\" data-testid=\\"new-expense__description\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"description\\"></label><label for=\\"new-expense__receipt-date\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Receipt date*</span><input required=\\"\\" id=\\"new-expense__receipt-date\\" data-testid=\\"new-expense__receipt-date\\" pattern=\\"\\\\d{4}-\\\\d{2}-\\\\d{2}\\" type=\\"date\\" min=\\"1900-01-01\\" max=\\"2077-01-01\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"2020-10-10\\"></label><label for=\\"new-expense__currency-select\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Currency*</span><select required=\\"\\" data-testid=\\"new-expense__currency-select\\" id=\\"new-expense__currency-select\\" class=\\"NewExpenseSelectField-jVTKdO iFJTld\\">
            <option value=\\"\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\"></option>
            <option value=\\"BRL\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">BRL</option>
            <option value=\\"USD\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">USD</option>
            <option value=\\"MXN\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">MXN</option>
          </select></label>
        <div class=\\"NewExpenseCurrencyInfoFieldContainier-buJYQd eLTHPp\\"><label for=\\"new-expense__receipt-value\\" width=\\"45%\\" class=\\"NewExpenseField-juMRoK jJGcov\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Receipt value*</span><input id=\\"new-expense__receipt-value\\" data-testid=\\"new-expense__receipt-value\\" type=\\"number\\" min=\\"0\\" required=\\"\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"\\"></label><label for=\\"new-expense__value-to-be-paid\\" width=\\"45%\\" class=\\"NewExpenseField-juMRoK jJGcov\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">To be paid value*</span><input id=\\"new-expense__value-to-be-paid\\" data-testid=\\"new-expense__value-to-be-paid\\" type=\\"number\\" min=\\"0\\" required=\\"\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"\\"></label></div>
      </div>
    </div>
    <div class=\\"NewExpenseFormActions-iMrQpP FOLRZ\\"><button id=\\"new-expense__btn-cancel\\" data-testid=\\"new-expense__btn-cancel\\" type=\\"button\\" class=\\"NewExpenseFormButton-eEGptI NewExpenseFormCancelButton-dqtNbK jrxbsp bYcGKK\\">Cancelar</button><button id=\\"new-expense__btn-submit\\" data-testid=\\"new-expense__btn-submit\\" type=\\"submit\\" class=\\"NewExpenseFormButton-eEGptI NewExpenseFormSaveButton-cHZoOV jrxbsp bTTZnZ\\">Salvar</button></div>
  </form>
</div>"
`;

exports[`NewExpenseComponent Snapshot tests should render new expense component without conditional fields 1`] = `
"<div class=\\"NewExpenseContainer-hqtosb juPdFW\\">
  <form class=\\"NewExpenseForm-cwuXvK kEHXvR\\"><span class=\\"NewExpenseTitle-lhcAMw jEIPWO\\">New expense</span>
    <div class=\\"NewExpenseContent-jatTpR kRaxfX\\">
      <div class=\\"NewExpenseFileContainer-bNEBXl cdSBvg\\">
        <div class=\\"NewExpenseFile-kZrDhS ldnNcp\\"><span class=\\"NewExpenseSpanFieldTitle-fryOLC eNcQVo\\">Receipt or invoice*</span><label for=\\"new-expense__file\\" class=\\"NewExpenseFileButton-doqqYw dmgrQy\\"><span>Select file</span><input id=\\"new-expense__file\\" data-testid=\\"new-expense__file\\" style=\\"display: none;\\" accept=\\".jpg,.png\\" type=\\"file\\"></label><span class=\\"NewExpenseHint-gvuMzb hbrupz\\">Just JPG or PNG images are accepted</span></div>
      </div>
      <div class=\\"NewExpenseFieldsContainer-eXgEkW kWXdsa\\"><label for=\\"new-expense__type-selec\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Type*</span><select id=\\"new-expense__type-select\\" data-testid=\\"new-expense__type-select\\" required=\\"\\" class=\\"NewExpenseSelectField-jVTKdO iFJTld\\">
            <option value=\\"\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\"></option>
            <option value=\\"hotel-fee\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">Hotel</option>
            <option value=\\"food\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">Food</option>
            <option value=\\"transport\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">Transport</option>
          </select></label><label for=\\"new-expense__description\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Expense description*</span><input required=\\"\\" id=\\"new-expense__description\\" data-testid=\\"new-expense__description\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"description\\"></label><label for=\\"new-expense__receipt-date\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Receipt date*</span><input required=\\"\\" id=\\"new-expense__receipt-date\\" data-testid=\\"new-expense__receipt-date\\" pattern=\\"\\\\d{4}-\\\\d{2}-\\\\d{2}\\" type=\\"date\\" min=\\"1900-01-01\\" max=\\"2077-01-01\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"2020-10-10\\"></label><label for=\\"new-expense__currency-select\\" class=\\"NewExpenseField-juMRoK hJPqOw\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Currency*</span><select required=\\"\\" data-testid=\\"new-expense__currency-select\\" id=\\"new-expense__currency-select\\" class=\\"NewExpenseSelectField-jVTKdO iFJTld\\">
            <option value=\\"\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\"></option>
            <option value=\\"BRL\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">BRL</option>
            <option value=\\"USD\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">USD</option>
            <option value=\\"MXN\\" class=\\"NewExpenseSelectOption-coaVTX eLEIqd\\">MXN</option>
          </select></label>
        <div class=\\"NewExpenseCurrencyInfoFieldContainier-buJYQd eLTHPp\\"><label for=\\"new-expense__receipt-value\\" width=\\"45%\\" class=\\"NewExpenseField-juMRoK jJGcov\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">Receipt value*</span><input id=\\"new-expense__receipt-value\\" data-testid=\\"new-expense__receipt-value\\" type=\\"number\\" min=\\"0\\" required=\\"\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"16\\"></label><label for=\\"new-expense__value-to-be-paid\\" width=\\"45%\\" class=\\"NewExpenseField-juMRoK jJGcov\\"><span class=\\"NewExpenseFieldLabel-iRoKGQ keXWMB\\">To be paid value*</span><input id=\\"new-expense__value-to-be-paid\\" data-testid=\\"new-expense__value-to-be-paid\\" type=\\"number\\" min=\\"0\\" required=\\"\\" class=\\"NewExpenseInputField-kvuItE eGrksb\\" value=\\"10\\"></label></div>
      </div>
    </div>
    <div class=\\"NewExpenseFormActions-iMrQpP FOLRZ\\"><button id=\\"new-expense__btn-cancel\\" data-testid=\\"new-expense__btn-cancel\\" type=\\"button\\" class=\\"NewExpenseFormButton-eEGptI NewExpenseFormCancelButton-dqtNbK jrxbsp bYcGKK\\">Cancelar</button><button id=\\"new-expense__btn-submit\\" data-testid=\\"new-expense__btn-submit\\" type=\\"submit\\" class=\\"NewExpenseFormButton-eEGptI NewExpenseFormSaveButton-cHZoOV jrxbsp bTTZnZ\\">Salvar</button></div>
  </form>
</div>"
`;
